diagnostic_aggregator/src/aggregator.cpp

```diff
@@ -107,13 +106,77 @@ void Aggregator::diagCallback(const diagnostic_msgs::DiagnosticArray::ConstPtr&
   }
 }
 
-Aggregator::~Aggregator() 
+Aggregator::~Aggregator()
 {
   if (analyzer_group_) delete analyzer_group_;
 
   if (other_analyzer_) delete other_analyzer_;
 }
 
+
+void Aggregator::bondBroken()
+{
+  boost::mutex::scoped_lock lock(mutex_); // Possibility of multiple bonds breaking at once
+
+  vector<boost::shared_ptr<bond::Bond> >::iterator it;
+  for (it = bonds_.begin(); it != bonds_.end();)
+  {
+    if ((*it)->isBroken())
+    {
+      string id = (*it)->getId();
+      ROS_DEBUG("bond for namespace %s was broken", id.c_str());
+      analyzer_group_->removeAnalyzer(added_analyzers_.find(id)->second);
+      added_analyzers_.erase(id);
+      bonds_.erase(it); // it points to next in list (or end())
+    } else {
+      // increment in afterthought will go past bonds_.end() if erase()ing last
+      // element, so do it here instead
+      ++it;
+    }
+  }
+}
+
+void Aggregator::bondFormed(){}
+
+bool Aggregator::addDiagnostics(diagnostic_msgs::AddDiagnostics::Request &req,
+				diagnostic_msgs::AddDiagnostics::Response &res)
+{
+  // Wait for connections on this bond. Wait until it's formed to put it in the
+  // bond vector
```

Comment:
Blocking here while waiting for the bond to form will block all processing of other service calls and message callbacks, including processing of messages on the diagnostics topic.

This setup should probably be done without blocking the rest of the ROS event loop, and it should probably include a timeout as well, in case the calling node dies and doesn't clean up properly (assuming that isn't already contained in the bond API)